### ISSUE TITLE
nzBeforeUpload from nz-upload should have the file parameter of type File instance instead of UploadFile

### ISSUE THREAD

**Issue by IonelLupu** [NG-ZORRO/ng-zorro-antd#4744]
### Reproduction link
[https://stackblitz.com/edit/ng-zorro-antd-start-ygithh?file=src/app/app.component.ts](https://stackblitz.com/edit/ng-zorro-antd-start-ygithh?file=src/app/app.component.ts)

### Steps to reproduce
Use a `nz-upload` with a `nzBeforeUpload` attribute that returns false. Log the `file` parameter inside `nzBeforeUpload`. 

### What is expected?
I expect the argument type of the `nzBeforeUpload` function to be of type File instead of UploadFile as the function actually has inside Javascript

### What is actually happening?
`nzBeforeUpload` has the wrong type definition for the file parameter of UploadFile instead of File.

| Environment | Info |
|---|---|
| ng-zorro-antd | 8.5.2 |
| Browser | * |

---
The type definition just needs to be changed to reflect what the function actually is called with: 
Now is:
```ts
nzBeforeUpload: (file: UploadFile, fileList: UploadFile[]) => boolean | Observable<boolean>;
```
It should be:
```ts
nzBeforeUpload: (file: File, fileList: File[]) => boolean | Observable<boolean>;
```

PS. I am not sure if the `fileList` parameter should be `File[]`.

**Comment by radiumrasheed** [comment]
the UploadFile already has the original file object as ```UploadFile.originFileObj```

**Comment by IonelLupu** [comment]
@radiumrasheed no. As seen in the demo, the `nzBeforeUpload` calls the method with a file of type `File` and the definition says it is of type `UploadFile`. There is an issue of definition on the `nzBeforeUpload` method inside nzUpload

**Comment by radiumrasheed** [comment]
actually yes, I agree to this. its really an issue. I noticed this after my comment

**Comment by emondora** [comment]
I agree to this.  In "beforeUpload", it is really a "File", not UploadFile. 
When I want to preview "base64 thumbnail" in nzBeforeUpload callback, I read the File Blob from UploadFile.originFileObj, but got undefined error! Because after "nzBeforeUpload", the uploadFile.originFileObj is not exist.
" beforeFrontUpload = (file: File) =>  { ...  } "  will solve the problem! But when you turn on "strictTemplates": true.
Everything Goes Wrong!

**Comment by adammruk** [comment]
I have exactly same case like @emondora. For now i use a small workaround:
In my `beforeUpload` function i accept file as UploadFile:
`
beforeUpload = (file: UploadFile): Observable<boolean> => {
`
but later i'm just casting types:
`
reader.readAsDataURL(file as unknown as File);
`

**Comment by IonelLupu** [comment]
This is still an issue in version `9.3.0`

**Comment by snebjorn** [comment]
I just encountered this as well.

https://stackblitz.com/edit/angular-y45bu7?file=src/app/app.component.ts

None of the `NzUploadFile` properties work and the type of the object [nzBeforeUpload] is called with is `File`

```ts
beforeUpload = (file: NzUploadFile, _fileList: NzUploadFile[]) => {
    console.log('Instance of: ' + file.constructor.name); // Instance of: File

    console.log(file.filename, file.originFileObj, file.error, file.linkProps, file.percent, file.thumbUrl, file.response, file.uid); 
   // undefined undefined undefined undefined undefined undefined undefined e630sazjkzv
}
```

**Comment by IonelLupu** [comment]
This is still an issue in version `10.0.0`

**Comment by cipchk** [comment]
@IonelLupu The type `file` is a primitive file type [File Object](https://developer.mozilla.org/en-US/docs/Web/API/File), This is an error on the documentation.

**Comment by IonelLupu** [comment]
@cipchk yes, but the `nzBeforeUpload` attribute expects a function which has the first parameter of type `NzUploadFile` and not `File`. So, the code needs to be updated as well. This is the definition of `nzBeforeUpload`:

```
nzBeforeUpload?: (file: NzUploadFile, fileList: NzUploadFile[]) => boolean | Observable<boolean>;
```

**Comment by snebjorn** [comment]
This is still an issue in v11.1.0.

Isn't this just a type correction? Though I guess this will be a breaking change for people using `strictTemplates`

**Comment by Linzeur** [comment]
Hi,

I am using the version 10.2.2 for my current project.

Days ago, I had the issue I wanted to get the value as datatype File from a variable NzUploadFile in the event beforeUpload because I need convert the file to base 64. However, when I used the property "originFileObj" from this variable but I always got undefined.

Then, I found a solution which I put here:

```
beforeUpload = (file: NzUploadFile): boolean => {
    
    this.fileList = [file];
    const myReader = new FileReader();
    myReader.readAsDataURL(file as any);
    myReader.onloadend = (e) => {
      this.base64 = myReader.result;
    };
    return false;
  };
```

The trick is pass the NzUploadFile variable in the method "readAsDataURL" change the datatype using the keyword "as". I was surprised when I tested that code and it worked. Although, I have doubts because it works, but now that was the solution which I reach.

**Comment by IonelLupu** [comment]
@Linzeur While this works, it is just a temporary solution. Using `as any` anywhere is problematic

**Comment by IonelLupu** [comment]
Bumping this a bit. I encountered this again.

**Comment by krokofant** [comment]
Still broken in version 13. 😮

**Comment by TIMPICKLE** [comment]
still broken in 13.2.2

**Comment by TIMPICKLE** [comment]
ahh..i thank it works. just receive a "File" type. and then use .arrayBuffer() to get a promise to process your data. 
``` 
    beforeUpload = async (file: File): Promise<boolean> => {
        file.arrayBuffer().then(async (data) => {
          // use your stream do anything
        }).catch(
          (error) => {console.log(error);
        })
        
        
        console.log(this.fileList);
        return false;
    };

**Comment by guoziyun** [comment]
just cast NzUploadFile to File type , it will work
```
beforeUpload = (file: NzUploadFile): boolean => {
    const formData = new FormData();
    formData.set("image", file as unknown as File);
}
```

**Comment by stljeff1** [comment]
bump. still broken in v14

**Comment by Nosfistis** [comment]
Actually, `NzUploadFile` should mention that it extends `File`. It would solve problems.

**Comment by snebjorn** [comment]
@Nosfistis as explained in https://github.com/NG-ZORRO/ng-zorro-antd/issues/4744#issuecomment-688710484 the parameter object (NzUploadFile) doesn't have any of the properties relevant to NzUploadFile set. They're all `undefined`.
So it's not just a matter of extending.

**Comment by Nosfistis** [comment]
@snebjorn it seems the only property that `NzUploadFile` actually attaches in `beforeUpload` is `uid`. The rest are probably for the automatic upload.
Having them `undefined` is not an issue as they are actually marked as optional. The actual implementation though does use `NzUploadFile` as an extension of `File`.

**Comment by jdonzallaz** [comment]
Still broken in v16

**Comment by JakobWl** [comment]
Still broken...

**Comment by linhnobi** [comment]
`<nz-upload
  [nzCustomRequest]="handleCustomUpload"
  [nzShowUploadList]="false"
  [nzAccept]="'.png,.jpg,.jpeg'"
  (nzChange)="handleChangeUpload($event)">
  <button nz-button>
    <i nz-icon nzType="upload"></i> Chọn ảnh
  </button>
</nz-upload>`


`handleCustomUpload = (item: NzUploadXHRArgs): Subscription => {
    // Check before processing
    if (!item.file) {
      console.error('item.file undefined!');
      return EMPTY.subscribe();
    }

    // Simulate successful upload
    setTimeout(() => {
      item.onSuccess?.({}, item.file, {});
    }, 500);

    return EMPTY.subscribe();
  };

  // ✅ Convert into base64 when file change
  async handleChangeUpload(info: NzUploadChangeParam): Promise<void> {
    const file = info.file;

    if (!file || !file.originFileObj) return;

    try {
      const base64 = await this.convertToBase64(file);
      this.imageBase64 = base64 as string;
    } catch (error) {
      console.error('Convert error:', error);
    }
  }

  // ✅ Convert file into base64
  convertToBase64(file: NzUploadFile): Promise<string | ArrayBuffer | null> {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      const originFile = file.originFileObj;

      if (!originFile) {
        reject('originFileObj does not exist');
        return;
      }

      reader.readAsDataURL(originFile);
      reader.onload = () => resolve(reader.result);
      reader.onerror = (error) => reject(error);
    });
  }`

I handle image uploading locally and convert the image to base64 format. From what I’ve learned, Ant Design uploads files to the server by default, so I modified the upload function to simulate a successful upload locally.

**Comment by RoroTitiFR** [comment]
I agree with this, this is really an issue... On my own, I did this workaround to work with File instead of "falsy" NzUploadFIle. $any() in the template is the trick to allow working with File without any compiling issue.

```ts
files = signal<File[]>([]);

beforeUpload = (file: File) => {
  this.files.set([file]);
  return false;
};
```

```html
<nz-upload nzType="drag" [(nzFileList)]="$any(files)" [nzBeforeUpload]="$any(beforeUpload)" [nzMultiple]="false">
  <p class="ant-upload-drag-icon">
    <nz-icon nzType="fa-regular:upload" />
  </p>
  <p class="ant-upload-text">
    Cliquez ou glissez dans cette zone le fichier CSV à importer
  </p>
</nz-upload>
```
